### PR TITLE
Feature/d3 asim 3589

### PR DESCRIPTION
**Pull request by spyrostz**
Added event dispatcher for the settlement markets. Adapted MycoMatcher to handle also settlement markets. Adapted the market classes to have dedicated debug info and a settlement market type.

### CODE REVIEW

File: src/d3a/models/market/two_sided.py
Comment (faizan2590) [inline src/d3a/models/market/two_sided.py:57]:
```suggestion
        return "TwoSidedMarket"
```

Not guaranteed to always be pay-as-bid at this point.

File: src/d3a/models/area/__init__.py
Comment (faizan2590) [inline src/d3a/models/area/__init__.py:352]: NIT: To avoid lazy imports

File: src/d3a/models/area/event_dispatcher.py
Comment (faizan2590) [inline src/d3a/models/area/event_dispatcher.py:52]:
```suggestion
        self._settlement_agents: Dict[DateTime, Dict[str, SettlementAgent]] = {}
```

File: src/d3a/models/area/market_rotators.py
Comment (faizan2590) [inline src/d3a/models/area/market_rotators.py:33]: Might better belongs to `d3a/src/d3a/models/market/__init__.py`

File: src/d3a/models/market/one_sided.py
Comment (faizan2590) [inline src/d3a/models/market/one_sided.py:64]:
```suggestion
        return "OneSidedMarket"
```

File: src/d3a/models/market/two_sided.py
Comment (faizan2590) [inline src/d3a/models/market/two_sided.py:400]: How about creating a module `src/d3a/models/market/settlement_market.py` and moving it there. Although now now `SettlementMarket ` is only a wrapper of `TwoSidedMarket `

File: src/d3a/models/market/two_sided.py
Comment (faizan2590) [inline src/d3a/models/market/two_sided.py:396]:
```suggestion
        return "SettlementMarket"
```

File: src/d3a/models/market/two_sided.py
Comment (spyrostz) [reply src/d3a/models/market/two_sided.py:400]: Works for me, will do so

File: src/d3a/models/area/market_rotators.py
Comment (spyrostz) [reply src/d3a/models/area/market_rotators.py:33]: Will move it thanks

File: src/d3a/models/strategy/external_strategies/pv.py
Comment (faizan2590) [inline src/d3a/models/strategy/external_strategies/pv.py:284]:
```suggestion
                               for v in market.get_offers().values()
```

File: tests/strategies/external/test_load.py
Comment (faizan2590) [inline tests/strategies/external/test_load.py:69]: I think `time_slot` being part of it too . Right?

Comment (spyrostz) [reply tests/strategies/external/test_load.py:69]: These are generic tests for the aggregator commands, I am not testing the time_slot functionality here. This is tested by test_get_market_from_cmd_arg_returns_settlement_market , test_get_market_from_cmd_arg_returns_spot_market tests

File: tests/strategies/settlement/test_settlement_strategy.py
Comment (faizan2590) [inline tests/strategies/settlement/test_settlement_strategy.py:93]: Sorry, didn't get the intuition of calling `event_tick ` twice

File: src/d3a/setup/settlement_market/settlement_market.py
Comment (faizan2590) [inline src/d3a/setup/settlement_market/settlement_market.py:29]: Is there any integration tests expected to use this setup file?

File: tests/strategies/settlement/test_settlement_strategy.py
Comment (spyrostz) [reply tests/strategies/settlement/test_settlement_strategy.py:93]: This is related to the change in the SettlementMarketStrategy.event_tick method. You will notice that the order of the offer/bid update and increment/update counter has been reversed (now the offer is updated before the counter is incremented). This is mandatory to have it in that order because AFAIR the  UpdateFrequencyMixin class works that way (Load and PV strategies follow the same logic, first they update their bids and offers, then they increment the counter of the UpdateFrequencyMixin). This has the effect that if the event_tick is only called once, then the bids and the offers of the strategies will not be updated (because the update counter will not have incremented yet). Therefore a second tick is really required in order for the bids and the offers to be updated with a new energy rate.

File: src/d3a/setup/settlement_market/settlement_market.py
Comment (spyrostz) [reply src/d3a/setup/settlement_market/settlement_market.py:29]: For now this works mostly as an example, but an integration test will follow soon

File: src/d3a/models/market/one_sided.py
Comment (hannesdiedrich) [inline src/d3a/models/market/one_sided.py:50]: Thanks a lot for refactoring this. I as also about to.

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: src/d3a/models/myco_matcher/myco_internal_matcher.py
Comment (spyrostz) [reply src/d3a/models/myco_matcher/myco_internal_matcher.py:61]: No, the settlement market will not be handled by Mycos ATM, and I am not sure if it ever will.

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:161]: This unfortunately is not feasible. This is a method of the AreaDispatcher class, therefore the AreaDispatcher class name is not visible from inside the class, thus the type hint will fail.

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:132]: First of all, double quotes does not remove the need to import the class of the type hint. Even in this case we need to do `from d3a.models.markets import Market` in order for python interpreter to not complain. As a followup to that, if I import the Area class in this module I will end up with a circular import error. I can avoid it by using this https://stackoverflow.com/a/39757388, but conditional imports only for type hints is very ugly IMO and IIRC we have discussed to not do it and simply not add type hints in these cases. TLDR, I will not add type hint for the owner argument

File: src/d3a/models/strategy/external_strategies/__init__.py
Comment (spyrostz) [reply src/d3a/models/strategy/external_strategies/__init__.py:353]: There is already such a property, `is_aggregator_controlled`. I'll reuse it

File: src/d3a/models/area/event_dispatcher.py
Comment (fievelk) [reply src/d3a/models/area/event_dispatcher.py:164]: I think using strings works in this case, as described in [PEP 484 (Forward references)](https://www.python.org/dev/peps/pep-0484/#forward-references)

File: src/d3a/models/area/event_dispatcher.py
Comment (fievelk) [reply src/d3a/models/area/event_dispatcher.py]: I haven't tested this, but [PEP 484 (Forward references)](https://www.python.org/dev/peps/pep-0484/#forward-references) describes a way to overcome circular imports:
> A common use for forward references is when e.g. Django models are needed in the signatures. Typically, each model is in a separate file, and has methods taking arguments whose type involves other models. Because of the way circular imports work in Python, it is often not possible to import all the needed models directly [...]
The solution is to switch to module-only imports and reference the models by their `_module_._class_ name`.

So for example:
```python
from models import b
class A(Model):
    def foo(self, b: 'b.B'):
        pass
```

File: src/d3a/models/area/event_dispatcher.py
Comment (fievelk) [reply src/d3a/models/area/event_dispatcher.py:164]: Another method would be to use:
```python
from __future__ import annotations
```
(Source: [How do I type hint a method with the type of the enclosing class?](https://stackoverflow.com/a/33533514/1916588))

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py]: I also do not like this solution actually, mainly due to the fact that module-based imports can be very broad, and also because they can lead to long/verbose types (e.g. one type hint could be d3a.models.area.Area which is overly verbose). More specifically in this example, this would not solve our problem because the Area class is on the d3a.models.area.__init__ module. Therefore when one does `from d3a.models import area` , the Area class will also be imported thus causing the circular import, despite the fact that area.Area would be used. 
The correct way to solve this is not by trying to cheat the Python interpreter, but by really refactoring the code in order to avoid such occurrences in the first place (by moving classes around files in a way that circular imports are avoided). However this is something that should be added as a separate task and not as part of this ticket.

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:166]: Indeed the forward reference worked here, fixing this

Comment (Hassan754) [reply src/d3a/models/area/event_dispatcher.py:166]: I'm not sure about this but I think that using a string typehint will raise a warning in pylint if the reference is not imported, this way the IDEs won't recognize the type, but we can live with that atm

Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:166]: The forward reference (double quotes) works for me, already tried it with pylint and does not complain.

Comment (faizan2590) [reply src/d3a/models/area/event_dispatcher.py:166]: This raises another question of why not changing this method from static_method to instance_method ?

**Review by faizan2590 — COMMENTED** [review]
Left one comment to a [thread](https://github.com/gridsingularity/d3a/pull/1221/files#r707225471), will LGTM once addressed/resolved.

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:166]: This is not called only for `self`, but also with `area.parent.event_dispatcher` as an argument, in order to connect 2 IAAs. Therefore it is not possible to change it to an instance method which implies that the argument will only be `self`

**Review by faizan2590 — DISMISSED** [review]
LGTM

File: src/d3a/d3a_core/redis_connections/redis_area_market_communicator.py
Comment (fievelk) [inline src/d3a/d3a_core/redis_connections/redis_area_market_communicator.py:107]: It looks like this method is never called. Is it a leftover?

File: src/d3a/d3a_core/util.py
Comment (fievelk) [inline src/d3a/d3a_core/util.py:174]: I suppose that SA stands for "Settlement Agent", right?
I would suggest moving this method inside the `SettlementAgent` class, as that's the only place where it is currently used.

File: src/d3a/models/area/__init__.py
Comment (fievelk) [inline src/d3a/models/area/__init__.py:356]: Nit: do we already have a ticket for this? Or should we create one?

File: src/d3a/models/area/markets.py
Comment (fievelk) [inline src/d3a/models/area/markets.py:108]: I really like this change 👍

File: src/d3a/d3a_core/util.py
Comment (spyrostz) [reply src/d3a/d3a_core/util.py:174]: Not a good idea, especially for the future when we add more integration tests for the settlement market. Check out the usages of the make_ba_name and make_iaa_name, they are also used in the context of integration tests.

File: src/d3a/d3a_core/redis_connections/redis_area_market_communicator.py
Comment (spyrostz) [reply src/d3a/d3a_core/redis_connections/redis_area_market_communicator.py:107]: Remnant after the merge from master, removing

File: src/d3a/models/area/__init__.py
Comment (spyrostz) [reply src/d3a/models/area/__init__.py:356]: I think we had one, but it is probably buried in Outdated. Creating one quickly

File: src/d3a/models/area/__init__.py
Comment (fievelk) [inline src/d3a/models/area/__init__.py:423]: Is there a specific reason to use `list` on this? The following should work as well:
```suggestion
                            "settlement_markets": self.settlement_markets.values(),
```

Comment (spyrostz) [reply src/d3a/models/area/__init__.py:423]: Should work fine even without the casting to list, removing

File: src/d3a/models/area/event_dispatcher.py
Comment (fievelk) [inline src/d3a/models/area/event_dispatcher.py:88]: Here we're accessing a private property of a different class (`area._markets`). We do the same in several other modules, so I think that `area._markets` wants to be made public.

File: src/d3a/models/area/event_dispatcher.py
Comment (fievelk) [inline src/d3a/models/area/event_dispatcher.py:101]: Nit:
```suggestion
        if not self.area.events.is_enabled and event_type not in [
                AreaEvent.ACTIVATE, AreaEvent.MARKET_CYCLE]:
            return
```

File: src/d3a/models/area/event_dispatcher.py
Comment (fievelk) [inline src/d3a/models/area/event_dispatcher.py:111]: Nit: in this class we have many `if` methods to basically distinguish between SPOT, BALANCING and SETTLEMENT market types (like [here](https://github.com/gridsingularity/d3a/blob/ea445ce2d33609c69a7dcd577f6978113475b137/src/d3a/models/area/event_dispatcher.py#L165-L174) and [here](https://github.com/gridsingularity/d3a/blob/ea445ce2d33609c69a7dcd577f6978113475b137/src/d3a/models/area/event_dispatcher.py#L144-L163)). 
I think that this could be refactored into separate classes (each specialized on one market). No need to do it in the context of this PR, but it might be good to keep it in mind and maybe open a refactoring ticket.

Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:111]: I would leave this for when we reimplement the event dispatcher to not forward the bids and the offers from the higher to the lower market, since this is going to be implemented in the near future (this is the last task of the degrees of freedom market)

File: src/d3a/models/area/redis_external_market_connection.py
Comment (fievelk) [inline src/d3a/models/area/redis_external_market_connection.py:156]: Nit:
```suggestion
        elif self.redis_communicator.is_enabled:
```
(Just for improved clarity)

File: src/d3a/models/market/one_sided.py
Comment (fievelk) [inline src/d3a/models/market/one_sided.py:66]: Nit: this could also be `self.__class__.__name__` (in case we want to make it dynamic)

File: src/d3a/models/market/one_sided.py
Comment (fievelk) [inline src/d3a/models/market/one_sided.py:307]: Do you confirm that the level of this is actually `info`? I'm asking because of the `_debug_log_market_type_identifier` variable.

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:88]: I would rather do this in isolation after this PR is merged, because this touches many parts of the code and is not related with the PR itself

File: src/d3a/models/area/event_dispatcher.py
Comment (spyrostz) [reply src/d3a/models/area/event_dispatcher.py:101]: Fixing

File: src/d3a/models/myco_matcher/myco_internal_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/myco_internal_matcher.py:60]:
```suggestion
        for area_data in self.area_uuid_markets_mapping.values():
```

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:156]: Fixing, will make it private too

File: src/d3a/models/market/one_sided.py
Comment (spyrostz) [reply src/d3a/models/market/one_sided.py:307]: It is info and it is important to stay that way. Otherwise you will not be able to see the trades in the default d3a log level

File: src/d3a/models/market/one_sided.py
Comment (spyrostz) [reply src/d3a/models/market/one_sided.py:66]: Will do and also remove this from the classes that inherit from this

File: src/d3a/models/myco_matcher/myco_internal_matcher.py
Comment (spyrostz) [reply src/d3a/models/myco_matcher/myco_internal_matcher.py:60]: Will fix this, however keep in mind that this might change again during this task https://gridsingularity.atlassian.net/browse/D3ASIM-3593

**Comment by fievelk** [comment]
General:
- Some parts of the new code are not unit-tested. Codecov is highlighting most of them, in case you want to have a look. 
  I personally believe that private methods should not be tested in general; however, I also have the feeling that some of these methods would actually be good candidates to become public methods of smaller classes (and therefore be tested) maybe in the context of a refactoring PR.

Nits:
- Please replace `\` with `()`
- Some missing docstrings for new classes and public methods

**Comment by spyrostz** [comment]
> General:
> 
>     * Some parts of the new code are not unit-tested. Codecov is highlighting most of them, in case you want to have a look.
>       I personally believe that private methods should not be tested in general; however, I also have the feeling that some of these methods would actually be good candidates to become public methods of smaller classes (and therefore be tested) maybe in the context of a refactoring PR.
> 
> 
> Nits:
> 
>     * Please replace `\` with `()`
> 
>     * Some missing docstrings for new classes and public methods

Regarding the \ , I found only one occurrence in the diff, which already existed there. Did you mean just that ? 
Also regarding the docstrings, really the only new classes are the SettlementMarket and SettlementAgent, both of which inherit from the TwoSided market and agent respectively without any new functionality apart from debug logs. IMO these do not really need docstrings at all. If you insist I will add these, however it is not helpful to my eyes. 

Regarding test coverage, I'll check it out to see exactly the places where github complains, however I doubt that this is due to this PR. E.g. one occurrence I saw from a first glance is the `match_recommendations` method, which was already there in the first place (all I did was to include the settlement markets in the matching) therefore won't be able to do much. I'll give it a closer look and let you know.

**Comment by fievelk** [comment]
> Regarding the \ , I found only one occurrence in the diff, which already existed there. Did you mean just that ?

Yes, I wrote a general comment as that's what we decided to do in the retrospective (for `'`, `\` and similar). Since it was just one case, I could have also just commented the line (but I didn't know yet while reviewing) :)

> Also regarding the docstrings, really the only new classes are the SettlementMarket and SettlementAgent, both of which inherit from the TwoSided market and agent respectively without any new functionality apart from debug logs. IMO these do not really need docstrings at all. If you insist I will add these, however it is not helpful to my eyes.

I wouldn't insist regarding this specific case, as the PR has already been thoroughly reviewed. But I would insist on it as a general process regarding PRs, because I think that docstrings are good to have, even when they seem redundant (as they are also used to generate automatic documentation); also, code without pylint warnings looks better to me 😄 

> Regarding test coverage, I'll check it out to see exactly the places where github complains, however I doubt that this is due to this PR. E.g. one occurrence I saw from a first glance is the `match_recommendations` method, which was already there in the first place (all I did was to include the settlement markets in the matching) therefore won't be able to do much. I'll give it a closer look and let you know.

This is actually a broader topic. I think that it would be nice for us to test methods even when they have not been implemented in our current PR. This is because we have a lot of existing untested code, and a sensible way to cover it is to gradually add tests for it when we modify it. Adding a line to an untested method without testing it would decrement our coverage even more.

**Comment by spyrostz** [comment]
> > Regarding the \ , I found only one occurrence in the diff, which already existed there. Did you mean just that ?
> 
> Yes, I wrote a general comment as that's what we decided to do in the retrospective (for `'`, `\` and similar). Since it was just one case, I could have also just commented the line (but I didn't know yet while reviewing) :)
> 
> > Also regarding the docstrings, really the only new classes are the SettlementMarket and SettlementAgent, both of which inherit from the TwoSided market and agent respectively without any new functionality apart from debug logs. IMO these do not really need docstrings at all. If you insist I will add these, however it is not helpful to my eyes.
> 
> I wouldn't insist regarding this specific case, as the PR has already been thoroughly reviewed. But I would insist on it as a general process regarding PRs, because I think that docstrings are good to have, even when they seem redundant (as they are also used to generate automatic documentation); also, code without pylint warnings looks better to me smile
> 

Ok, since it helps you I'll add a couple of lines for each. However IMO we need to rethink this policy (I am not sure even if we have one for this). We have discussed it in the past too, but docstrings are not always good to have, even if automatic documentation is generated, because they are prone to diverge from the actual code. And especially for self-documenting classes that are almost dummy, I doubt that the comment serves more functionality other than satisfying pylint to be honest (speaking of, we can disable this in pylintrc, and I would vote in favor to do so)

> > Regarding test coverage, I'll check it out to see exactly the places where github complains, however I doubt that this is due to this PR. E.g. one occurrence I saw from a first glance is the `match_recommendations` method, which was already there in the first place (all I did was to include the settlement markets in the matching) therefore won't be able to do much. I'll give it a closer look and let you know.
> 
> This is actually a broader topic. I think that it would be nice for us to test methods even when they have not been implemented in our current PR. This is because we have a lot of existing untested code, and a sensible way to cover it is to gradually add tests for it when we modify it. Adding a line to an untested method without testing it would decrement our coverage even more.

It actually depends on the PR if you ask me. For this specific PR, where it touches almost all the tricky parts of the code (MycoInternalMatcher, AreaMarkets, EventDispatcher are some of these, that do not have dedicated tests but are tested through other classes' tests) one cannot really expect that all these classes should be unit tested in one PR, especially since there are integration tests in the PR that already cover the basic functionality.

**Comment by fievelk** [comment]
> Ok, since it helps you I'll add a couple of lines for each. However IMO we need to rethink this policy (I am not sure even if we have one for this). We have discussed it in the past too, but docstrings are not always good to have, even if automatic documentation is generated, because they are prone to diverge from the actual code. And especially for self-documenting classes that are almost dummy, I doubt that the comment serves more functionality other than satisfying pylint to be honest (speaking of, we can disable this in pylintrc, and I would vote in favor to do so)

I agree with you on the fact that in some cases docstrings are not so necessary. However, I think this is the minority of the cases, and I'd rather have redundant docstrings enforced by pylint than not having important ones because we disabled the check. If a class is trivial, you can skip reading its docstring while you write new code. If a class/method is complex, you want to have a good docstring for it (and it will save you a lot of time). Of course, having confusing/outdated docstrings can be worse, but that's part of the tasks of a PR review IMO.

> It actually depends on the PR if you ask me. For this specific PR, where it touches almost all the tricky parts of the code (MycoInternalMatcher, AreaMarkets, EventDispatcher are some of these, that do not have dedicated tests but are tested through other classes' tests) one cannot really expect that this PR has to add dedicated unit tests for all these, especially since there are integration tests in the PR that already cover the basic functionality. Keep in mind that this PR is not a refactoring task, but is dealing with a task/feature that needs to be delivered in the short term.

Definitely: adding all unit-tests in one go for all these classes would not be feasible. As we discussed in the retrospective, all parts of the implementation process (including tests and conversation with stakeholders) should be considered when determining the complexity of new tasks. For settlement markets this is more complicated, and that's also why I added the comment as a "nit" that can be disregarded.

**Review by fievelk — DISMISSED** [review]

File: src/d3a/d3a_core/util.py
Comment (fievelk) [reply src/d3a/d3a_core/util.py:170]: Thanks @spyrostz. Your answer got me thinking so I ask here, even though the PR has been merged. I have a doubt: when should these integration tests fail? For example, suppose I change the implementation of `make_ba_name` to be something like:
```python
def make_ba_name(owner):
    return "Some wrong string or value"
```
Should end-to-end tests fail?
I personally think they should, because one "side" is expecting to receive `f"BA {owner.name}"`, but it's receiving something else. If this is the case, then it would be good to directly use the expected string in integration tests, without calling `make_ba_name()`. Using the existing implementation is tempting, because we wouldn't have to manually write the string; but it actually will not allow us to be alerted if the implementation is incorrect.

What is your opinion on this?

**Pull request merged**
